### DISCUSSION TITLE
Polybar on headless remote host complains `No monitors found"

### DISCUSSION THREAD

**Discussion by anoduck** [polybar/polybar#2948] [Q&A] [answered]
OS: Kali Linux Rolling amd64
Window Manager: Bspwm
Remote Technology: TigerVNC
Polybar Version: 3.6.3

Trying to run polybar on a headless server remotely, I am receiving the message `no monitors found`. 

The output of `xrandr --query` displays the available monitors correctly, as follows:

```bash
Screen 0: minimum 320 x 200, current 1916 x 1041, maximum 8192 x 8192
VGA-0 disconnected primary (normal left inverted right x axis y axis)
DVI-0 disconnected (normal left inverted right x axis y axis)
```

Yet, `polybar -m` displays nothing and cleanly exists, and so does `polybar -M`. I have tried manually setting the variable in the configuration file to `Screen 0`, `VGA-0, and `DVI-0`. I have tried the rather creative launch script, but no progress has been made. 

The only solution I can think of is creating a `fake monitor` in order to "trick" polybar into working. Maybe something with Xvfb (?).

**Comment by patrick96** [comment, marked as answer]
The problem here is that the monitors are disconnected (can be ignored with `monitor-strict = false`) and don't have any size and position information associated with them (cannot be ignored).

Polybar needs the position information to position itself and the height and width information to support percentages in the `width` and `height` settings.

So, a fake monitor is also the only solution I could think of here. Unless you can somehow get your VNC server to create a connected monitor when you connect.

Note: There have been requests before to allow polybar to run without specifying a monitor by manually setting its position, but no one has worked on this, see #2456.

**Reply by anoduck** [reply]
You are very agreeable, did you know that? I appreciate the info. 

> So, a fake monitor is also the only solution I could think of here. Unless you can somehow get your VNC server to create a connected monitor when you connect.

Hit the nail on the head with that one. I will look into what other VNC servers provide. I am currently running Tiger's implementation of Xvnc, but I do remember there are some differences in X11vnc. I am working on it.

**Comment by anoduck** [comment]
Well, that worked, presented a new issue for my specific instance, but it worked. Here is the rundown:

I created a fake monitor using the `xserver-xorg-video-dummy` package, then I created a new xorg configuration, finally I configured lightdm to launch `X` with the manually created `xorg.conf`.

```config
Section "Monitor"
  Identifier "Monitor0"
  HorizSync 28.0-80.0
  VertRefresh 48.0-75.0
  # https://arachnoid.com/modelines/
  # 1920x1080 @ 60.00 Hz (GTF) hsync: 67.08 kHz; pclk: 172.80 MHz
  Modeline "1920x1080_60.00" 172.80 1920 2040 2248 2576 1080 1081 1084 1118
EndSection

Section "Device"
  Identifier "Dummy_Card"
  Driver "dummy"
  VideoRam 256000
EndSection

Section "Screen"
  DefaultDepth 24
  Identifier "Dummy_Screen"
  Device "Dummy_Card"
  Monitor "Dummy_Monitor"
  SubSection "Display"
    Depth 24
    Modes "1920x1080_60.00"
  EndSubSection
EndSection
```

The only drawback to this is where before Xvnc would allow my monitor resolution to expand and contract dynamically, it now is set to an annoyingly small `1024x800`.

**Reply by patrick96** [reply]
Nice! I'm sure this is also going to be helpful for other people trying to use polybar over VNC

**Comment by anoduck** [comment]
Just an update to the dummy configuration. After a refresher on `xorg.conf`, I tweaked the configuration, which provided a much better experience. Bear with me, I have not manually had to configure `xorg.conf` in quite a while.

```config
Section "Screen"
  Identifier	"Dummy_Screen"
  Device	"Dummy_Card"
  Monitor	"Dummy_Monitor"
  DefaultDepth	24
  SubSection	"Display"
    Depth	24
  EndSubSection
EndSection

Section "Device"
  Identifier	"Dummy_Card"
  Driver 	"dummy"
  OPTION	"DPMS"
EndSection

Section "Monitor"
  Identifier	"Monitor0"
  HorizSync 	30.0-83.0
  VertRefresh 	56.0-76.0
  # https://arachnoid.com/modelines/
  # 1920x1080 @ 60.00 Hz (GTF) hsync: 67.08 kHz; pclk: 172.80 MHz
  Modeline "1920x1080_60.00" 172.80 1920 2040 2248 2576 1080 1081 1084 1118 -HSync +Vsync
  Option	"DPMS"
EndSection
```
